MeetupPlatformApi/Controllers/AuthenticationController.cs

```diff
@@ -46,34 +46,99 @@ public async Task<IActionResult> RegisterUser([FromBody] UserRegistrationDto reg
         var userInfoDto = mapper.Map<UserOutputDto>(user);
         var outputDto = new UserRegistrationResultDto
         {
             UserInfo = userInfoDto,
-            AccessToken = accessToken
+            TokenPair = new() { AccessToken = tokenPair.AccessToken, RefreshTokenId = tokenPair.RefreshToken.Id }
         };
         return CreatedAtAction(nameof(GetUserById), new {id = outputDto.UserInfo.Id}, outputDto);
     }
 
+    [HttpPost("refreshTokens/{id:guid}")]
+    public async Task<IActionResult> TokenRefresh(Guid id)
+    {
+        var refreshToken = await context.RefreshTokens.Where(refreshToken => refreshToken.Id == id).SingleOrDefaultAsync();
+
+        if(refreshToken is null)
+        {
+            return BadRequest($"Token with id: {id} doesn't exist.");
+        }
+
+        var user = await context.Users.Where(user => user.Id == refreshToken.UserId).SingleOrDefaultAsync();
+
+        var tokenPair = authenticationManager.IssueTokenPair(user);
+
+        context.RefreshTokens.Remove(refreshToken);
+        await context.RefreshTokens.AddAsync(tokenPair.RefreshToken);
+        await context.SaveChangesAsync();
+
+        var outputDto = new AuthenticationTokenPairOutputDto { AccessToken = tokenPair.AccessToken, RefreshTokenId = tokenPair.RefreshToken.Id };
+        return Ok(outputDto);
+    }
+
+    [HttpDelete("{userId:guid}/refreshTokens")]
+    public async Task<IActionResult> RevokeAllRefreshTokens(Guid userId)
+    {
+        var user = await context.Users.Where(user => user.Id == userId).SingleOrDefaultAsync();
+
+        if (user is null)
+        {
+            return BadRequest($"User with id: {userId} doesn't exist.");
+        }
+
+        var userRefreshTokens = await context.RefreshTokens.Where(token => token.UserId == userId).ToListAsync();
+
+        context.RefreshTokens.RemoveRange(userRefreshTokens);
+        await context.SaveChangesAsync();
+        return NoContent();
+    }
+
     [HttpPost("authenticate")]
     public async Task<IActionResult> Authenticate([FromBody] UserAuthenticationDto authenticationDto)
     {
         var user = await context.Users.SingleOrDefaultAsync(user => user.Username == authenticationDto.Username);
+
         if (user is null || !BCrypt.Verify(authenticationDto.Password, user.Password))
         {
             return BadRequest("Username or password is incorrect.");
         }
 
-        var accessToken = authenticationManager.IssueAccessToken(user);
-        var outputDto = new AuthenticationTokenOutputDto { AccessToken = accessToken };
+        var tokenPair = authenticationManager.IssueTokenPair(user);
+
+        await context.RefreshTokens.AddAsync(tokenPair.RefreshToken);
+        await context.SaveChangesAsync();
+
+        var outputDto = new AuthenticationTokenPairOutputDto { AccessToken = tokenPair.AccessToken, RefreshTokenId = tokenPair.RefreshToken.Id };
         return Ok(outputDto);
     }
 
+    [HttpPut]
```

Comment:
We are not updating the whole user, but only his credentials, so the url should be either `/api/users/{id}/credentials` (and we should specify user's id) or `/api/users/me/credentials` (and requester should be authentication).